src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -27,144 +32,190 @@ public static partial class Base64
-                Unsafe.Add(ref destBytes, destIndex + 1) = (byte)(i0 >> 8);
-                destIndex += 2;
-            }
-            else
+            fixed (byte* srcBytes = &MemoryMarshal.GetReference(utf8))
+            fixed (byte* destBytes = &MemoryMarshal.GetReference(bytes))
             {
-                if (i0 < 0)
-                    goto InvalidExit;
-                if (destIndex > destLength - 1)
-                    goto DestinationSmallExit;
-                Unsafe.Add(ref destBytes, destIndex) = (byte)(i0 >> 16);
-                destIndex += 1;
-            }
-
-            sourceIndex += 4;
-
-            if (srcLength != utf8.Length)
-                goto InvalidExit;
+                int srcLength = utf8.Length & ~0x3;  // only decode input up to the closest multiple of 4.
+                int destLength = bytes.Length;
+                int maxSrcLength = srcLength;
+                int decodedLength = GetMaxDecodedFromUtf8Length(srcLength);
+
+                // max. 2 padding chars
+                if (destLength < decodedLength - 2)
+                {
+                    // For overflow see comment below
+                    maxSrcLength = destLength / 3 * 4;
+                }
+
+                byte* src = srcBytes;
+                byte* dest = destBytes;
+                byte* srcEnd = srcBytes + (uint)srcLength;
+                byte* srcMax = srcBytes + (uint)maxSrcLength;
+
+                if (maxSrcLength >= 24)
+                {
+                    byte* end = srcMax - 45;
+                    if (Avx2.IsSupported && (end >= src))
+                    {
+                        Avx2Decode(ref src, ref dest, end, maxSrcLength, destLength, srcBytes, destBytes);
+
+                        if (src == srcEnd)
+                            goto DoneExit;
+                    }
+
+                    end = srcMax - 24;
+                    if (Ssse3.IsSupported && (end >= src))
+                    {
+                        Ssse3Decode(ref src, ref dest, end, maxSrcLength, destLength, srcBytes, destBytes);
+
+                        if (src == srcEnd)
+                            goto DoneExit;
+                    }
+                }
+
+                // Last bytes could have padding characters, so process them separately and treat them as valid only if isFinalBlock is true
+                // if isFinalBlock is false, padding characters are considered invalid
+                int skipLastChunk = isFinalBlock ? 4 : 0;
+
+                if (destLength >= decodedLength)
+                {
+                    maxSrcLength = srcLength - skipLastChunk;
+                }
+                else
+                {
+                    // This should never overflow since destLength here is less than int.MaxValue / 4 * 3 (i.e. 1610612733)
+                    // Therefore, (destLength / 3) * 4 will always be less than 2147483641
+                    maxSrcLength = (destLength / 3) * 4;
```

Comment:
Should probably add an assert validating the assumption above?

Reply:
Added in 26fbd9348ac3389df35151c5b079cbde4549b24a